pkg/lexer/lexer.go

```diff
@@ -0,0 +1,259 @@
+package lexer
+
+import (
+	"unicode"
+	"unicode/utf8"
+)
+
+type TokenType int
+
+// These are the main token types for shell
+const (
+	// Special tokens
+	ILLEGAL TokenType = iota
+	SEMICOLON
+	GREATER
+	LESS
+	GREATGREAT
+	LESSLESS
+	GREATAMPERSAND
+	LESSAMPERSAND
+	PIPE
+	AMPERSAND
+	IDENT
+	SINGLEQUOTE
+	DOUBLEQUOTE
+	BACKQUOTE
+	COMMENT
+	EOF
+)
+
+type Token struct {
+	Type TokenType
+	Val  string
+}
+
+// grammar
+// args -> IDENTS | empty
+// cmd and args -> IDENT args
+// pipe list -> PIPE cmd and args | cmd and args
+// io modifier -> GREAT IDENT | GREATGREAT IDENT | ...
+// io modifiers list -> io modifiers | empty
+// background optional -> ampersand | empty
+// cmd line -> pipe list io modifiers background optional NEWLINE | NEWLINE | ILLEGAL recovery.
+
+type Lexer struct {
+	input  string // input string
+	start  int    // start position of current token
+	pos    int    // current position in input
+	width  int    // width of the last rune read
+	Tokens chan Token
+}
+
+func Lex(input string) *Lexer {
+	l := &Lexer{
+		input:  input,
+		Tokens: make(chan Token),
+	}
+	go l.run()
+	return l
+}
+
+func (l *Lexer) run() {
+	for state := lexCommand; state != nil; {
+		state = state(l)
+	}
+	close(l.Tokens)
+}
+
+func (l *Lexer) emit(t TokenType) {
+	l.Tokens <- Token{t, l.input[l.start:l.pos]}
+	l.start = l.pos
+}
+
+func (l *Lexer) next() rune {
+	if l.pos >= len(l.input) {
+		l.width = 0
+		return -1 // EOF
+	}
+	r, w := utf8.DecodeRuneInString(l.input[l.pos:])
+	l.width = w
+	l.pos += l.width
+	return r
+}
+
+func (l *Lexer) peek() rune {
```

Comment:
Since you are using `peek` inside `next`, you will need to change this code to:
```go
func (l *Lexer) peek() rune, width {
  if l.pos >= len(l.input) {
    l.width = 0
    return -1
  }
  return utf8.DecodeRuneInString(l.input[l.pos:])
}
```

Reply:
`peek` is not being used inside next though

Reply:
No, but I suggested you to use it: https://github.com/raklaptudirm/mash/pull/3/files#r752033616

Reply:
That creates an additional overhead in the state machine. I kept this lexer very function and made it so that I don't have state that can be inferred directly by the function - in this case ch can easily be returned by the function and I don't see why that is an issue? Is there a specific reason?

The main reason for my reluctance is because having it returned means that I can easily create and use separate functions that are not doing anything to the state of the lexer but still be able to make concrete checks. Plus, having it as a `.ch` makes the switch statement in `lexCommand` makes this messy (IMO) and as you can see the function `isIdentable` is leveraging the value (I agree that you can make the argument in switch statement be `l.ch` but the `l.next` is much cleaner IMO.

Reply:
Since there is no major issues with your implementation, you can continue with the current way of returning the value. It should be easy to change it in the future if required.